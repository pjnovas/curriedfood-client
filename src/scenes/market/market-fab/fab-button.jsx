import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet } from 'react-native';
import { FAB } from 'react-native-paper';

import { composeHooks } from 'utils/language';
import ConfirmDialog from './confirm-dialog';

export const FabButton = ({ ingredients, visible, onDismiss, onPress }) => {
  if (!ingredients.length) return null;

  return visible ? (
    <ConfirmDialog {...{ ingredients, onDismiss }} />
  ) : (
    <FAB style={styles.fab} icon="cart-plus" onPress={onPress} />
  );
};

FabButton.propTypes = {
  ingredients: PropTypes.array,
  visible: PropTypes.bool,
  onDismiss: PropTypes.func,
  onPress: PropTypes.func
};

FabButton.defaultProps = {
  ingredients: [],
  visible: false
};

const styles = StyleSheet.create({
  fab: {
    position: 'absolute',
    margin: 16,
    right: 0,
    bottom: 0
  }
});

export const useConfirm = () => {
  const [visible, setVisible] = useState(false);

  return {
    visible,
    onDismiss: () => setVisible(false),
    onPress: () => setVisible(true)
  };
};

export default composeHooks({ useConfirm })(FabButton);
