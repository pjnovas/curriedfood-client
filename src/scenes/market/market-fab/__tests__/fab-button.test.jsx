import * as React from 'react';
import { FabButton, useConfirm } from '../fab-button';

const ingredients = [
  {
    id: 1,
    quantity: 3,
    product: {
      name: 'Eggs',
      divisible: false,
      unit: 'un'
    }
  },
  {
    id: 2,
    quantity: 150,
    product: {
      name: 'Olives',
      divisible: true,
      unit: 'gr'
    }
  }
];

takeSnapshots(
  {
    default: {},
    'with ingredients': {
      ingredients
    },
    'ConfirmDialog when visible': {
      ingredients,
      visible: true
    }
  },
  FabButton
);

describe('useConfirm', () => {
  let spyUseState;
  const setVisible = jest.fn();
  const mockMarket = (visible) => {
    React.useState.mockImplementation(() => [visible, setVisible]);
  };

  beforeAll(() => {
    spyUseState = jest.spyOn(React, 'useState');
  });

  afterAll(() => {
    spyUseState.mockReset();
  });

  beforeEach(() => {
    setVisible.mockClear();
    React.useState.mockReset();
  });

  test('return a visible state', () => {
    mockMarket(true);
    expect(useConfirm()).toEqual(
      expect.objectContaining({
        visible: true
      })
    );

    mockMarket(false);
    expect(useConfirm()).toEqual(
      expect.objectContaining({
        visible: false
      })
    );
  });

  test('onDismiss allows to hide', () => {
    mockMarket();
    expect(setVisible).not.toBeCalled();
    useConfirm().onDismiss();
    expect(setVisible).toBeCalledWith(false);
  });

  test('onPress allows to show', () => {
    mockMarket();
    expect(setVisible).not.toBeCalled();
    useConfirm().onPress();
    expect(setVisible).toBeCalledWith(true);
  });
});
