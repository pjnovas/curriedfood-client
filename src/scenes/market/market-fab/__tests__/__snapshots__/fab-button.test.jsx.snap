// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog when visible 1`] = `
<Component
  ingredients={
    Array [
      Object {
        "id": 1,
        "product": Object {
          "divisible": false,
          "name": "Eggs",
          "unit": "un",
        },
        "quantity": 3,
      },
      Object {
        "id": 2,
        "product": Object {
          "divisible": true,
          "name": "Olives",
          "unit": "gr",
        },
        "quantity": 150,
      },
    ]
  }
/>
`;

exports[`default 1`] = `""`;

exports[`with ingredients 1`] = `
<withTheme(FAB)
  icon="cart-plus"
  style={
    Object {
      "bottom": 0,
      "margin": 16,
      "position": "absolute",
      "right": 0,
    }
  }
/>
`;
